### ISSUE TITLE
Style checker does not find missing spaces after commas in template lists

### ISSUE THREAD

**Issue by UnorderedSigh** [endless-sky/endless-sky#7954]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

This should trigger errors from the style checker:

```
template <class K,class V>
class whatever
{
public:
   whatever();
};

template <class K,class V>
whatever<K,V>::whatever()
{
}
```

### Steps to Reproduce

1. Put code like that in the repository.
2. Do a PR.
3. Look at the style checker results.
4. Sadness.

### Expected Behavior

All lines without a ` ` after a `,` in template lists should get style errors since the `<   >` act as parentheses in those situations, and commas in parenthetic lists should have spaces after them.

### Operating System

`Linux 5.0.0-37-generic #40~18.04.1-Ubuntu SMP Thu Nov 14 12:06:39 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`

### Game Source

Built from source

### Game Version

57d30ea from PR #7947

### Additional Information

@Hurleveur discovered this error in an earlier version of #7947 (hash [57d30ea](https://github.com/endless-sky/endless-sky/pull/7947/commits/57d30ea0602974df82a3ee8ac33dd9a7d0bf2e63))

**Comment by Hurleveur** [comment]
@tibetiroka I think this should be an easy fix? Just expect a space after a comma?

**Comment by tibetiroka** [comment]
@Hurleveur @UnorderedSigh I wish it were that easy. It came up in the early stages of development, and the answer is unforntunately that it isn't convenient to fix.

Currently all segments with type parameters is ignored by the style checker. This is a fix for the roughly 2.8k errors that would pop up due to the missing whitespaces around the `<` and `>` characters, the <> includes, random `*`-s for pointer types, etc.

It is pretty much impossible to rework all regex checks to work around these issues. I tried, and it didn't end well. That's why these were excluded in the first place.
<details><summary>Details</summary>
Include statements are not much of an issue, but type names and variables names are pretty hard to distinguish via regex. Even if we maintain a list of "primitive" type names, assume that all other names start with a capital letter and all variable names start with a lowercase, we have to account for any references to variables in main.cpp (although that really, really shouldn't happen) because that would also start with a lowercase letter.  Oh, and it sometimes starts with `class` or other bullshit that is lowercase.
<br><br>
Types and comparisons are so close text-wise that it is sometimes nearly impossible to distinguish between them, at least without needing to go through a significant part of the C++ standard.

Ideas are welcome, but please keep in mind that any more 'just make this a special case' kind of excludes would probably not get merged, we have too many of those already (two, to be exact).

</details>


Commas are searched outside of type parameters, so it works for e.g. method parameters. This just ended up as one of the edge cases that the style checker doesn't cover.

**Comment by UnorderedSigh** [comment]
Could it be handled by requiring a space after *any* comma outside of strings and comments?  Or am I missing something?  I guess there's the special case of `operator,` but it's generally unwise to override the `,` operator.

**Comment by tibetiroka** [comment]
That already is the case, but certain segments of the code are excluded from style checks. Any 'segment' (part of a line between strings and comments) with a `<` followed by a `>` is not checked. Overall, this is a small part of the code, but in this case it kind of messes everything up.

Again, you are welcome to look at the style checker and think about some other solution, but I don't want to add a new level of exclusions/checks to go around this issue.

**Comment by UnorderedSigh** [comment]
It sounds like the problem is that you're trying to use regular expressions to parse a context-sensitive language, which is provably impossible (it would violate the Pumping Lemma). Why don't you use a C++ style checking program instead?

**Comment by tibetiroka** [comment]
It doesn't use "proper" lints because the available solutions (such as clang-format) are not flexible enough / cannot be made to match our formatting rules so closely. Also, it would be a tremendous effort to create and maintain a fully-featured lint; certainly not something I could do in what little free time I have.

That's why we (well, I) decided to cheat, and create a bunch of regexes that match malformatted code with pretty good accuracy. It doesn't even attempt to parse C++ - in fact, that is exactly what it tries to avoid as much as possible.

This tool is not a replacement for code reviews, not even for the coding style part, but it helps reviewers by filtering out a bunch of issues before anyone even looks at the PR.

**Comment by UnorderedSigh** [comment]
It sounds like the solution, in an imaginary universe with infinite labor available, is:

1. A regular expression-based system to handle things regular linters can't.
2. A fine-tuned, purpose-built, C++ linter.
3. Possibly some minor updates to the code to match the new C++ linter's requirements, if we end up with extra requirements as a result of the linter's limitations.

However, lacking infinite labor, we just have to work around the limitations of the existing system.

**Comment by Hurleveur** [comment]
I think this just is overkill, we don't really need the style checker to check for everything.

**Comment by UnorderedSigh** [comment]
No, not everything, but it should do a better job of templates. You can't handle templates with regular expressions. That requires a proper C++ linter.

**Comment by Amazinite** [comment]
> I think this just is overkill, we don't really need the style checker to check for everything.

Yeah, this definitely doesn't seem like something worth pursuing a fix for, especially since it's such a rare occurrence that we have code like this anyway, and it's a simple fix to mention in a review.
